source/grid/tria_description.cc

```diff
@@ -0,0 +1,662 @@
+      const unsigned int my_rank_in)
+    {
+      if (auto tria_pdt = dynamic_cast<
+            const parallel::distributed::Triangulation<dim, spacedim> *>(&tria))
+        AssertThrow(comm == tria_pdt->get_communicator(),
+                    ExcMessage("MPI communicators do not match."));
+
+      // First, figure out for what rank we are supposed to build the
+      // ConstructionData object
+      unsigned int my_rank = my_rank_in;
+      AssertThrow(my_rank == numbers::invalid_unsigned_int ||
+                    my_rank < dealii::Utilities::MPI::n_mpi_processes(comm),
+                  ExcMessage(
+                    "Rank has to be smaller than available processes."));
+
+      if (auto tria_pdt = dynamic_cast<
+            const parallel::distributed::Triangulation<dim, spacedim> *>(&tria))
+        {
+          AssertThrow(
+            my_rank == numbers::invalid_unsigned_int ||
+              my_rank == dealii::Utilities::MPI::this_mpi_process(comm),
+            ExcMessage(
+              "If parallel::distributed::Triangulation as source triangulation, my_rank has to equal global rank."));
+
+          my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+        }
+      else if (auto tria_serial =
+                 dynamic_cast<const dealii::Triangulation<dim, spacedim> *>(
+                   &tria))
+        {
+          if (my_rank == numbers::invalid_unsigned_int)
+            my_rank = dealii::Utilities::MPI::this_mpi_process(comm);
+        }
+      else
+        {
+          AssertThrow(
+            false, ExcMessage("This type of triangulation is not supported!"));
+        }
+
+      Description<dim, spacedim> construction_data;
+
+      // store the communicator
+      construction_data.comm = comm;
+
+
+      std::map<unsigned int, std::vector<unsigned int>>
+                                           coinciding_vertex_groups;
+      std::map<unsigned int, unsigned int> vertex_to_coinciding_vertex_group;
+
+      GridTools::collect_coinciding_vertices(tria,
+                                             coinciding_vertex_groups,
+                                             vertex_to_coinciding_vertex_group);
+
+      // helper function, which collects all vertices belonging to a cell
+      // (also taking into account periodicity)
+      auto add_vertices_of_cell_to_vertices_owned_by_locally_owned_cells =
+        [coinciding_vertex_groups, vertex_to_coinciding_vertex_group](
+          TriaIterator<CellAccessor<dim, spacedim>> &cell,
+          std::vector<bool> &vertices_owned_by_locally_owned_cells) {
+          // add local vertices
+          for (unsigned int v = 0; v < GeometryInfo<dim>::vertices_per_cell;
+               ++v)
+            {
+              vertices_owned_by_locally_owned_cells[cell->vertex_index(v)] =
+                true;
+              const auto coinciding_vertex_group =
+                vertex_to_coinciding_vertex_group.find(cell->vertex_index(v));
```

Comment:
The map `vertex_to_coinciding_vertex_group` should be quite small since it only contains vertices coinciding due to periodic boundary conditions. 

I have run the function for large triangulations. See https://github.com/dealii/dealii/issues/8558#issuecomment-532250763